### PR TITLE
Leaves - Hallie - Media Ranker

### PR DESCRIPTION
**Pull request by idhallie**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote a custom model method to sort works by category and order them by number of votes.
Describe how you approached testing that model method. What edge cases did you come up with? | I test sorting by 'book' and check that the count matches what's in my fixtures and that the first and last item is a book. I also test sorting by movie, which has zero instances to ensure that it returns a count of 0 and that the first element is nil (which is redundant if the count is zero, but you know, I did it anyway).
What are `session` and `flash`? What is the difference between them? | Session is a built-in hash where you can store data from one request to be accessed later (ex. user information when logged in). Flash stores information temporarily; either for a single request cycle or within the same request (ex. flash.now). Flash is handy for issuing messages to the user (ex. validation errors on form entries or page access restrictions). 
What was one thing that you gained more clarity on through this assignment? | I gained more clarity on using fixtures and how to access information from related models. I created a handy counter-cache in my works model to make it easy to order works by number of votes. I'm pleased with how that works.
What is the Heroku URL of your deployed application? | https://media-ranker-by-hallie.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | I enjoyed this project and don't really have suggestions for the project itself. However, regarding relevant lectures, I thought it could have been helpful to have more model test and application helper examples. Admittedly, I did not use application helper because it seemed easier to write the small code snippet to manipulate the date than to use the application helper and test it.

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  
+  def logout
+    session[:user_id] = nil
+    flash[:success] = "Successfully logged out."
+    redirect_to root_path
+  end
+  
+  private
+  
+  def find_user
```
Comment (beccaelenzil) [inline app/controllers/users_controller.rb:52]: If you put this in the ApplicationController other controllers will inherit it.

**Comment by beccaelenzil** [comment]
# Media Ranker
## What We're Looking For



### Manual Testing

Workflow | yes / no
--- | ---
Deployed to Heroku | yes
Optional: Look and feel is similar to the original (consider styling and layout) | yes
**Before logging in** |
On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | yes
Can go into a work's show page | yes
Verify unable to vote on a work, and get a flash message | yes
Can edit this work successfully, and get a flash message | yes
Can go to "View all media" page and see three lists of works, sorted by vote | yes
Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | yes
Can create a new work successfully. Note the URL for this work's show page | yes
Can delete this work successfully |yes
Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | yes
Verify that the "View all users" page lists no users (except what is produced by db seeding) | yes
**Log in** |
Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | yes
Your username is listed in "View all users" page | yes
Verify that number of votes determines the Media Spotlight | yes
Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | yes
Voting on the same work twice produces an error/flash message, and there is no extra vote | yes
**Log out** |
Logging out showed a flash message and changed the UI | yes
Logging in as a new user creates a new user | yes
Logging in as an already existing user has a specific flash message |yes 

### Targeted Code Review

Area | yes / no
--- | ---
git commits were small and atomic, with useful messages | yes
Routes file uses `resources` for works |yes
Routes file shows intention in limiting routes for voting, log-in functionality, and users | yes
The homepage view, all media view, and new works view use semantic HTML | yes
The homepage view, all media view, and new works view use partials when appropriate | I see you used iteration to DRY up your code. That work! :)
The model for media (likely named `work.rb`) `has_many` votes | yes
The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | yes
Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user |see comment 
Some controller, likely the `WorksController`, has a controller filter for finding a work | yes
The `WorksController` uses strong params | yes
The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | yes

#### Targed Test Review

Area | yes / no
--- | ---
There are valid fixtures files used for users, votes, and works | yes
User model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | yes
Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Work model has tests with a section on all business logic methods in the model, including their edge cases |yes 

## Overall Feedback
Great work! Your code is well written and well tested. The learning goals for this project were clearly met. Keep up the hard work!
